compact.go

```diff
@@ -844,31 +839,107 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 			}
 			valset.set(l.Value)
 		}
-		postings.Add(i, lset)
-
-		i++
 	}
 	if set.Err() != nil {
 		return errors.Wrap(set.Err(), "iterate compaction set")
 	}
 
-	s := make([]string, 0, 256)
+	if meta.Stats.NumSamples == 0 {
+		// No postings to write, exit early.
+		return nil
+	}
+
+	return c.writePostings(indexw, values, set.seriesMap, indexReaders)
+}
+
+// writePostings writes the postings into the index file.
+func (c *LeveledCompactor) writePostings(indexw IndexWriter, values map[string]stringset,
+	seriesMap []map[uint64]uint64, indexReaders []IndexReader) (err error) {
+	var (
+		maxNumValues      int
+		numLabelValues    int
+		names             = make([]string, 0, len(values)+1)
+		apkName, apkValue = index.AllPostingsKey()
+	)
+	for _, v := range values {
+		numLabelValues += len(v)
+		if len(v) > maxNumValues {
+			maxNumValues = len(v)
+		}
+	}
+
+	labelValuesBuf := make([]string, 0, maxNumValues)
 	for n, v := range values {
-		s = s[:0]
+		labelValuesBuf = labelValuesBuf[:0]
+		names = append(names, n)
 
-		for x := range v {
-			s = append(s, x)
+		for val := range v {
+			labelValuesBuf = append(labelValuesBuf, val)
 		}
-		if err := indexw.WriteLabelIndex([]string{n}, s); err != nil {
+		if err := indexw.WriteLabelIndex([]string{n}, labelValuesBuf); err != nil {
 			return errors.Wrap(err, "write label index")
 		}
 	}
+	names = append(names, apkName)
+	sort.Strings(names)
+
+	if idxw, ok := indexw.(*index.Writer); ok {
+		idxw.HintPostingsWriteCount(numLabelValues)
+	}
 
-	for _, l := range postings.SortedKeys() {
-		if err := indexw.WritePostings(l.Name, l.Value, postings.Get(l.Name, l.Value)); err != nil {
-			return errors.Wrap(err, "write postings")
+	postingBuf := make([]uint64, 0, 1000000)
+	var bigEndianPost index.Postings = index.NewBigEndianPostings(nil)
```

Comment:
Can we make `NewBigEndianPostings` return `index.Postings`?

Reply:
This is contrary to what you said for `ListPostings`. I will make `NewListPostings` returning `ListPostings` and keep this as it is for consistency.

Reply:
Well it's useless to return an internal type when calling an exported function. Any caller must know it's an index.Postings and cast it to that. Exported functions should return exported types :)

Reply:
> Exported functions should return exported types

Now, this makes sense to me.